lib/rewrite/rewrite-set-facility.c

```diff
@@ -0,0 +1,144 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "rewrite-set-facility.h"
+#include "template/templates.h"
+#include "syslog-names.h"
+#include "scratch-buffers.h"
+#include "parse-number.h"
+
+#include <stdlib.h>
+
+typedef struct _LogRewriteSetFacility LogRewriteSetFacility;
+
+struct _LogRewriteSetFacility
+{
+  LogRewrite super;
+  LogTemplate *facility;
+};
+
+static gint
+_convert_facility_as_number(const gchar *facility_text)
+{
+  gint64 facility = 0;
+
+  if (!parse_dec_number(facility_text, &facility))
+    return -1;
+
+  if (facility > 127)
+    return -1;
+
+  return FACILITY_CODE(facility);
+}
+
+static gint
+_convert_facility_as_text(const gchar *facility_text)
+{
+  return syslog_name_lookup_facility_by_name(facility_text);
+}
+
+static gint
+_convert_facility(const gchar *facility_text)
+{
+  gint facility = _convert_facility_as_number(facility_text);
+  if (facility >= 0)
+    return facility;
+
+  facility = _convert_facility_as_text(facility_text);
+  if (facility >= 0)
+    return facility;
+
+  return -1;
+}
+
+static void
+_set_msg_facility(LogMessage *msg, const guint32 facility)
+{
+  msg->pri = (msg->pri & ~LOG_FACMASK) | facility;
+}
+
+static void
+log_rewrite_set_facility_process(LogRewrite *s, LogMessage **pmsg, const LogPathOptions *path_options)
+{
+  ScratchBuffersMarker marker;
+  GString *result = scratch_buffers_alloc_and_mark(&marker);
+  LogRewriteSetFacility *self = (LogRewriteSetFacility *) s;
+
+  log_msg_make_writable(pmsg, path_options);
+
+  log_template_format(self->facility, *pmsg, NULL, LTZ_LOCAL, 0, NULL, result);
+
+  const gint facility = _convert_facility(result->str);
+  if (facility < 0)
+    {
+      msg_debug("Warning: invalid value passed to set-facility()",
+                evt_tag_str("facility", result->str),
+                log_pipe_location_tag(&s->super));
+      goto error;
+    }
+
+  msg_trace("Setting syslog facility",
+            evt_tag_int("old_facility", (*pmsg)->pri & LOG_FACMASK),
+            evt_tag_int("new_facility", facility),
+            evt_tag_printf("msg", "%p", *pmsg));
+  _set_msg_facility(*pmsg, facility);
+
+error:
+  scratch_buffers_reclaim_marked(marker);
+}
+
+static LogPipe *
+log_rewrite_set_facility_clone(LogPipe *s)
+{
+  LogRewriteSetFacility *self = (LogRewriteSetFacility *) s;
+  LogRewriteSetFacility *cloned = (LogRewriteSetFacility *)log_rewrite_set_facility_new(log_template_ref(self->facility),
+                                  s->cfg);
+
+  if (self->super.condition)
+    cloned->super.condition = filter_expr_ref(self->super.condition);
```

Comment:
`rewrite_condition_opt` is missing from grammar.

Reply:
Thanks for spotting this. Probably it is missing in set-severity () too, as I was borrowing code from that.